### PR TITLE
wgengine/monitor: parse Linux netlink messages, ignore our own events

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes tailscale/corp#412 ("flood of link change events at start-up")

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/528)

### CODE REVIEW

File: wgengine/monitor/monitor.go [hunk 1]
```diff
@@ -140,3 +143,42 @@ func (m *Mon) debounce() {
...
+// newAddrMessage is a message for a new address being added.
+type newAddrMessage struct {
+	Addr  netaddr.IP
+	Label string // netlink Label attribute on Linux (e.g. "tailscale0")
+}
+
+var cgnatPrefix, _ = netaddr.ParseIPPrefix("100.64.0.0/10")
+
+func (m *newAddrMessage) ignore() bool {
+	return cgnatPrefix.Contains(m.Addr)
```
Comment (danderson) [inline wgengine/monitor/monitor.go:183]: Same here, don't ignore the ChromeOS range.

File: wgengine/monitor/monitor_linux.go [hunk 2]
```diff
@@ -44,12 +50,58 @@ func (c *nlConn) Close() error {
...
+			return nil, nil
+		}
+	}
+	msg := c.buffered[0]
+	c.buffered = c.buffered[1:]
+
+	// See https://github.com/torvalds/linux/blob/master/include/uapi/linux/rtnetlink.h
+	// And https://man7.org/linux/man-pages/man7/rtnetlink.7.html
+	const (
+		RTM_NEWADDR  = 20
```
Comment (danderson) [inline wgengine/monitor/monitor_linux.go:71]: There are probably in x/sys/unix if you wanted to do that?

File: wgengine/monitor/monitor.go [hunk 3]
```diff
@@ -140,3 +143,42 @@ func (m *Mon) debounce() {
 		}
 	}
 }
+
+// message represents a message returned from an osMon.
+//
+// Currently those are newRouteMessage and newAddrMessage.
+//
+// For historical reasons, a nil value means "some unspecified change
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply wgengine/monitor/monitor.go:151]: I was too lazy to update FreeBSD. But done now. Removed this comment.

File: wgengine/monitor/monitor.go [hunk 4]
```diff
@@ -140,3 +143,42 @@ func (m *Mon) debounce() {
...
+// you should react to and not ignore".
+type message interface {
+	// Ignore is whether we should ignore this message.
+	ignore() bool
+}
+
+// newRouteMessage is a message for a new route being added.
+type newRouteMessage struct {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply wgengine/monitor/monitor.go:159]: Done.

File: wgengine/monitor/monitor.go [hunk 5]
```diff
@@ -140,3 +143,42 @@ func (m *Mon) debounce() {
...
+	Src, Dst, Gateway netaddr.IP
+	Table             uint8 // table number ofor OSes that have one
+}
+
+func (m *newRouteMessage) ignore() bool {
+	if runtime.GOOS == "linux" && m.Table == 88 {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply wgengine/monitor/monitor.go:165]: done

File: wgengine/monitor/monitor.go [hunk 6]
```diff
@@ -140,3 +143,42 @@ func (m *Mon) debounce() {
...
+		return true
+	}
+	if cgnatPrefix.Contains(m.Dst) {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply wgengine/monitor/monitor.go:168]: net/interfaces.go already has:

```go
func IsTailscaleIP(ip net.IP) bool {
        nip, _ := netaddr.FromStdIP(ip) // TODO: push this up to caller, change func signature                                                                   
        return cgNAT.Contains(nip)
}
```

... which is used by tsshd at least.

I'll clean all this up first in a separate CL.

File: wgengine/monitor/monitor_linux.go (hunk 2 above)
Comment (bradfitz) [reply wgengine/monitor/monitor_linux.go:71]: good call

File: wgengine/monitor/monitor.go (hunk 1 above)
Comment (bradfitz) [reply wgengine/monitor/monitor.go]: Done. (see new tsaddr package)

**Pull request merged**